File: cloud-storage-spring-java/src/main/java/com/github/sevtech/cloud/storage/spring/config/GoogleCloudStorageConfig.java
package com.github.sevtech.cloud.storage.spring.config;

import com.github.sevtech.cloud.storage.spring.property.GoogleCloudStorageProperties;
import com.github.sevtech.cloud.storage.spring.service.StorageService;
import com.github.sevtech.cloud.storage.spring.service.impl.GoogleCloudStorageService;
import com.google.auth.oauth2.ServiceAccountCredentials;
import com.google.cloud.storage.Storage;
import com.google.cloud.storage.StorageOptions;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.io.FileInputStream;
import java.io.IOException;

@Slf4j
@Configuration
@ConditionalOnCloudStorageProperty(value = "gcp.storage.enabled")
public class GoogleCloudStorageConfig {

    @Bean
    public GoogleCloudStorageProperties googleCloudStorageProperties() {
        return new GoogleCloudStorageProperties();
    }

    @Bean
    public Storage storageClient(final GoogleCloudStorageProperties googleCloudStorageProperties) throws IOException {
        log.info("Registering Google Storage client");
        return StorageOptions.newBuilder()
                .setCredentials(ServiceAccountCredentials.fromStream(new FileInputStream(googleCloudStorageProperties.getKeyfile())))
                .build()
                .getService();
    }

    @Bean
    public StorageService googleCloudStorageService(final Storage storageClient) {
        return new GoogleCloudStorageService(storageClient);
    }

}

File: cloud-storage-spring-kotlin/src/main/kotlin/com/github/sevtech/cloud/storage/spring/config/GoogleCloudStorageConfig.kt
package com.github.sevtech.cloud.storage.spring.config

import com.github.sevtech.cloud.storage.spring.property.GoogleCloudStorageProperties
import com.github.sevtech.cloud.storage.spring.service.StorageService
import com.github.sevtech.cloud.storage.spring.service.impl.GoogleCloudStorageService
import com.google.auth.oauth2.ServiceAccountCredentials
import com.google.cloud.storage.Storage
import com.google.cloud.storage.StorageOptions
import mu.KotlinLogging
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import java.io.FileInputStream
import java.io.IOException

@Configuration
@ConditionalOnCloudStorageProperty(value = "gcp.storage.enabled")
class GoogleCloudStorageConfig {

    private val log = KotlinLogging.logger {}

    @Bean
    fun googleCloudStorageProperties(): GoogleCloudStorageProperties? {
        return GoogleCloudStorageProperties()
    }

    @Bean
    @Throws(IOException::class)
    fun storageClient(googleCloudStorageProperties: GoogleCloudStorageProperties): Storage? {
        log.info("Registering Google Storage client")
        return StorageOptions.newBuilder()
            .setCredentials(ServiceAccountCredentials.fromStream(FileInputStream(googleCloudStorageProperties.keyfile!!)))
            .build()
            .service
    }

    @Bean
    fun googleCloudStorageService(storageClient: Storage): StorageService {
        return GoogleCloudStorageService(storageClient)
    }

}